### PR TITLE
Allow changing the weight given to classifying conditions

### PR DESCRIPTION
**Pull request by mariushoch**
Allow changing the weight given to classifying conditions via the `$wgPropertySuggesterClassifyingConditionWeight` setting. This only applies to the "item" context if there is at least one classifying statement.

During testing it turned out that the classifying conditions resulted in way better suggestions than the ones which are purely based on whether a Statement with a given Property id exists. Due to this, I want to be able to boost the weight of these classifying conditions, in order to address T132839.

https://phabricator.wikimedia.org/T132839

### CODE REVIEW

File: src/PropertySuggester/Suggesters/SimpleSuggester.php [hunk 1]
```diff
@@ -73,38 +78,58 @@ public function setInitialSuggestions( array $initialSuggestionIds ) {
...
+		$hasClassifyingPropertyIds = false;
+		foreach ( $idTuples as $tuple ) {
+			$tupleConditions[] = $this->buildTupleCondition( $tuple[0], $tuple[1] );
+			$hasClassifyingPropertyIds = $hasClassifyingPropertyIds || $tuple[1] > 0;
 		}
-		else{
-			$condition = $dbr->makeList( $idTuples, LIST_OR );
+
+		$dbr = $this->lb->getConnection( DB_REPLICA );
+		if ( $hasClassifyingPropertyIds && $this->classifyingConditionWeight != 0.5 ) {
```
Comment (filbertkm) [inline src/PropertySuggester/Suggesters/SimpleSuggester.php]: why != instead of !== ?

Comment (mariushoch) [inline src/PropertySuggester/Suggesters/SimpleSuggester.php]: I forgot about this… I actually wanted to replace this with a safer float comparison.

**Review by addshore — APPROVED** [review]

File: src/PropertySuggester/Suggesters/SimpleSuggester.php [hunk 2]
```diff
@@ -114,14 +141,39 @@ protected function getSuggestions( array $propertyIds, array $idTuples, $limit,
...
 		return $this->buildResult( $res );
 	}
 
 	/**
+	 * @param string $dbType
+	 * @param int $count
+	 * @return string
+	 */
+	private function getWeightedPropSelect( $dbType, $count ) {
+		// The sum of the weights needs to be 2.
```
Comment (thiemowmde) [inline src/PropertySuggester/Suggesters/SimpleSuggester.php]: Why does it need to be 2? I believe I understood, but the comment should explain.

File: src/PropertySuggester/Suggesters/SimpleSuggester.php [hunk 3]
```diff
@@ -73,38 +78,60 @@ public function setInitialSuggestions( array $initialSuggestionIds ) {
...
+	 * @param array[] $idTuples array of ( int Property id, int Item id) tuples
 	 * @param int $limit
 	 * @param float $minProbability
 	 * @param string $context
+	 *
 	 * @throws InvalidArgumentException
+	 *
 	 * @return Suggestion[]
 	 */
-	protected function getSuggestions( array $propertyIds, array $idTuples, $limit, $minProbability, $context ) {
```
Comment (thiemowmde) [inline src/PropertySuggester/Suggesters/SimpleSuggester.php]: It looks like the removal of this parameter could easily be a separate patch. Splitting this into more smaller patches would help a lot in getting reviews for this significant change. Same for the removal of `protected`. I'm willing to help and upload such patches, if there is a chance they get merged fast.

File: README.md [hunk 4]
```diff
@@ -42,9 +42,13 @@ generate this data from a wikidata dump.
 * $wgPropertySuggesterMinProbability - a float that sets a minimum threshold for suggestions (default 0.05)
 * $wgPropertySuggesterDeprecatedIds - a list of ints that blacklist suggestions
 * $wgPropertySuggesterInitialSuggestions - a list of ints that will be suggested when no statements exist
+* $wgPropertySuggesterClassifyingConditionWeight - a float between 0 and 1 that determines the weight that's given to classifying property ids relative to non classifying ones (default 0.5)
 
 ## Release notes
 
+### 4.0.0 (dev)
```
Comment (thiemowmde) [inline README.md]: I wonder why this additional setting requires such a major version bump? As far as I can tell this is not a breaking change.

**Review by thiemowmde — APPROVED** [review]
I do have quite some suggestions, but at the same time I think we should move forward, merge this, experiment and allow us to possibly fail. It's a setting. The theory is that we can simply change this setting back to the default, and everything will be back to as it was before. Right?

**Comment by mariushoch** [comment]
Please note: This is not yet ready to be merged: There is still an open design question which I need to properly encode in configuration or decide (with the community and the PM) before merging.

**Comment by thiemowmde** [comment]
I manually rebased this now after #192 got merged.

**Comment by Ladsgroup** [comment]
@mariushoch Can you move this PR to gerrit? It would be great.

**Comment by Ladsgroup** [comment]
I'm closing this PR as this repo is moved to gerrit (https://gerrit.wikimedia.org/r/#/q/project:mediawiki/extensions/PropertySuggester) and it's WIP and not ready for review (yet). I won't delete this branch (and no one should) to make this easy to pick up and move forward.
